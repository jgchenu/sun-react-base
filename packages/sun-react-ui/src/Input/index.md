## Input

### 默认输入框

```tsx
import { Input } from 'sun-react-ui';
import React from 'react';

export default () => (
  <Input
    style={{ width: 200 }}
    placeholder="placeholder"
    onChange={console.log}
  />
);
```

### 受控输入框

```tsx
import { Input } from 'sun-react-ui';
import React, { useState } from 'react';

export default () => {
  const [value, setValue] = useState('');
  return (
    <Input
      value={value}
      defaultValue={value}
      onChange={(e) => {
        setValue(e.target.value);
      }}
    />
  );
};
```

### 禁用输入框

```tsx
import { Input } from 'sun-react-ui';
import React from 'react';

export default () => (
  <Input style={{ width: 300 }} placeholder="disabled input" disabled />
);
```

### 图标输入框

```tsx
import { Input } from 'sun-react-ui';
import { Search } from 'sun-react-icons';

import React from 'react';

export default () => (
  <Input
    style={{ width: 300 }}
    placeholder="input with icon"
    icon={<Search />}
  />
);
```

### 前缀跟后缀输入框

```tsx
import { Input } from 'sun-react-ui';

import React from 'react';

export default () => (
  <div>
    <Input
      style={{ width: 300 }}
      defaultValue="prepend text"
      prepend="https://"
    />
    <Input style={{ width: '300px' }} defaultValue="google" append=".com" />
  </div>
);
```

### 不同大小的输入框

```tsx
import { Input } from 'sun-react-ui';

import React from 'react';

export default () => (
  <div>
    <Input style={{ width: 300 }} defaultValue="large size" size="large" />
    <Input style={{ width: 300 }} placeholder="small size" size="small" />
  </div>
);
```
