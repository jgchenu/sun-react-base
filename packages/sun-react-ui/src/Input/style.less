@import '../styles/index';

@input-prefix-cls: ~'@{sun-prefix}-input';

@input-inner-cls: ~'@{input-prefix-cls}-inner';

.@{input-prefix-cls} {
  &-wrap {
    position: relative;
    display: flex;
    width: 100%;
  }

  &-icon-wrap {
    position: absolute;
    top: 0;
    right: 0;
    display: flex;
    width: 35px;
    height: 100%;
    align-items: center;
    justify-content: center;
    color: @input-color;

    svg {
      color: @input-placeholder-color;
    }

    + .@{input-inner-cls} {
      padding-right: 35px;
    }
  }

  &-inner {
    width: 100%;
    padding: @input-padding-y @input-padding-x;
    border: @input-border-width solid @input-border-color;
    background-clip: padding-box;
    background-color: @input-bg;
    border-radius: @input-border-radius;
    color: @input-color;
    font-family: @input-font-family;
    font-size: @input-font-size;
    font-weight: @input-font-weight;
    line-height: @input-line-height;
    transition: @input-transition;

    &:hover {
      border-color: fade(@input-focus-border-color, 70);
    }

    &:focus {
      border-color: @input-focus-border-color;
      background-color: @input-focus-bg;
      box-shadow: @input-focus-box-shadow;
      color: @input-focus-color;
      outline: 0;
    }

    &::placeholder {
      color: @input-placeholder-color;
      opacity: 1;
    }

    &:disabled,
    &[readonly] {
      border-color: @input-disabled-border-color;
      background-color: @input-disabled-bg;
      opacity: 1;
    }
  }

  &-small-size {
    .@{input-inner-cls} {
      padding: @input-padding-y-sm @input-padding-x-sm;
      border-radius: @input-border-radius-sm;
      font-size: @input-font-size-sm;
    }
  }

  &-large-size {
    .@{input-inner-cls} {
      padding: @input-padding-y-lg @input-padding-x-lg;
      border-radius: @input-border-radius-lg;
      font-size: @input-font-size-lg;
    }
  }

  &-prepend,
  &-append {
    display: flex;
    align-items: center;
    padding: @input-padding-y @input-padding-x;
    border: @input-border-width solid @input-group-addon-border-color;
    margin-bottom: 0; // Allow use of <label> elements by overriding our default margin-bottom
    background-color: @input-group-addon-bg;
    border-radius: @input-border-radius;
    color: @input-group-addon-color;
    font-size: @input-font-size; // Match inputs
    font-weight: @font-weight-normal;
    line-height: @input-line-height;
    text-align: center;
    white-space: nowrap;
  }

  // clear both radius
  &-append {
    .border-left-radius(0);
  }

  &-has-append {
    .@{input-inner-cls} {
      .border-right-radius(0);
    }
  }

  // clear both radius
  &-prepend {
    .border-right-radius(0);
  }

  &-has-prepend {
    .@{input-inner-cls} {
      .border-left-radius(0);
    }
  }
}
