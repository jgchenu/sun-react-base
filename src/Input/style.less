@import './../styles/index';

.sun-input-wrap {
  position: relative;
  display: flex;
  width: 100%;
  margin-bottom: 30px;

  .icon-wrap {
    position: absolute;
    top: 0;
    right: 0;
    display: flex;
    width: 35px;
    height: 100%;
    align-items: center;
    justify-content: center;
    color: @input-color;

    svg {
      color: @input-placeholder-color;
    }
  }
}

.sun-input-inner {
  width: 100%;
  padding: @input-padding-y @input-padding-x;
  border: @input-border-width solid @input-border-color;
  background-clip: padding-box;
  background-color: @input-bg;
  border-radius: @input-border-radius;
  color: @input-color;
  font-family: @input-font-family;
  font-size: @input-font-size;
  font-weight: @input-font-weight;
  line-height: @input-line-height;
  transition: @input-transition;

  &:hover {
    border-color: fade(@input-focus-border-color, 70);
  }

  &:focus {
    border-color: @input-focus-border-color;
    background-color: @input-focus-bg;
    box-shadow: @input-focus-box-shadow;
    color: @input-focus-color;
    outline: 0;
  }

  &::placeholder {
    color: @input-placeholder-color;
    opacity: 1;
  }

  &:disabled,
  &[readonly] {
    border-color: @input-disabled-border-color;
    background-color: @input-disabled-bg;
    opacity: 1;
  }
}

.icon-wrap + .sun-input-inner {
  padding-right: 35px;
}

.sun-input-group-prepend,
.sun-input-group-append {
  display: flex;
  align-items: center;
  padding: @input-padding-y @input-padding-x;
  border: @input-border-width solid @input-group-addon-border-color;
  margin-bottom: 0; // Allow use of <label> elements by overriding our default margin-bottom
  background-color: @input-group-addon-bg;
  border-radius: @input-border-radius;
  color: @input-group-addon-color;
  font-size: @input-font-size; // Match inputs
  font-weight: @font-weight-normal;
  line-height: @input-line-height;
  text-align: center;
  white-space: nowrap;
}

.input-size-small .sun-input-inner {
  padding: @input-padding-y-sm @input-padding-x-sm;
  border-radius: @input-border-radius-sm;
  font-size: @input-font-size-sm;
}

.input-size-large .sun-input-inner {
  padding: @input-padding-y-lg @input-padding-x-lg;
  border-radius: @input-border-radius-lg;
  font-size: @input-font-size-lg;
}

.input-group > .sun-input-group-prepend,
.input-group.input-group-append > .sun-input-inner {
  .border-right-radius(0);
}

.input-group > .sun-input-group-append,
.input-group.input-group-prepend > .sun-input-inner {
  .border-left-radius(0);
}
