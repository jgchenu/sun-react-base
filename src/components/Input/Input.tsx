import React, { FC, ReactNode, InputHTMLAttributes, ChangeEvent } from "react";
import classNames from "classnames";

type InputSize = "lg" | "sm";
export interface BasicInputProps {
  /**是否禁用 Input */
  disabled?: boolean;
  /**设置 input 大小，支持 lg 或者是 sm */
  size?: InputSize;
  /**添加图标，在右侧悬浮添加一个图标，用于提示 */
  icon?: ReactNode;
  /**添加前缀 用于配置一些固定组合 */
  prepend?: string | ReactNode;
  /**添加后缀 用于配置一些固定组合 */
  append?: string | ReactNode;
  onChange?: (e: ChangeEvent<HTMLInputElement>) => void;
}

type InputProps = BasicInputProps &
  Omit<InputHTMLAttributes<HTMLInputElement>, "size" | "onChange" | "disabled">;

const Input: FC<InputProps> = (props) => {
  const { disabled, size, icon, prepend, append, style, ...restProps } = props;
  const cnames = classNames("sun-input-wrap", {
    [`input-size-${size}`]: !!size,
    "is-disabled": !!disabled,
    "input-group": !!prepend || !!append,
    "input-group-append": !!append,
    "input-group-prepend": !!prepend,
  });
  const fixControlledValue = (value: any) => {
    if (typeof value === "undefined" || value === null) {
      return "";
    }
    return value;
  };
  if ("value" in props) {
    delete restProps.defaultValue;
    restProps.value = fixControlledValue(props.value);
  }
  return (
    <div className={cnames} style={style}>
      {prepend && <div className="sun-input-group-prepend">{prepend}</div>}
      {icon && <div className="icon-wrap">{icon}</div>}
      <input className="sun-input-inner" disabled={disabled} {...restProps} />
      {append && <div className="sun-input-group-append">{append}</div>}
    </div>
  );
};

export default Input;
