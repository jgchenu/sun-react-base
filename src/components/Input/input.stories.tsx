import React, { useState } from "react";
import { storiesOf } from "@storybook/react";
import { action } from "@storybook/addon-actions";
import { Input } from "./input";
import { FaSearch } from "react-icons/fa";
const ControlledInput = () => {
  const [value, setValue] = useState();
  return (
    <Input
      value={value}
      defaultValue={value}
      onChange={(e) => {
        setValue(e.target.value as any);
      }}
    />
  );
};
const defaultInput = () => (
  <section>
    <Input
      style={{ width: "300px" }}
      placeholder="placeholder"
      onChange={action("changed")}
    />
    <ControlledInput />
  </section>
);
const disabledInput = () => (
  <Input style={{ width: "300px" }} placeholder="disabled input" disabled />
);

const iconInput = () => (
  <Input
    style={{ width: "300px" }}
    placeholder="input with icon"
    icon={<FaSearch />}
  />
);

const sizeInput = () => (
  <section>
    <Input style={{ width: "300px" }} defaultValue="large size" size="lg" />
    <Input style={{ width: "300px" }} placeholder="small size" size="sm" />
  </section>
);

const pandInput = () => (
  <section>
    <Input
      style={{ width: "300px" }}
      defaultValue="prepend text"
      prepend="https://"
    />
    <Input style={{ width: "300px" }} defaultValue="google" append=".com" />
  </section>
);

storiesOf("Input component", module)
  .add("Input", defaultInput)
  .add("被禁用的 Input", disabledInput)
  .add("带图标的 Input", iconInput)
  .add("大小不同的 Input", sizeInput)
  .add("带前后缀的 Input", pandInput);
