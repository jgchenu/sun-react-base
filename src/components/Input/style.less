.sun-input-wrap {
  position: relative;
  display: flex;
  width: 100%;
  margin-bottom: 30px;

  .icon-wrap {
    position: absolute;
    top: 0;
    right: 0;
    display: flex;
    width: 35px;
    height: 100%;
    align-items: center;
    justify-content: center;
    color: @input-color;

    svg {
      color: @input-placeholder-color;
    }
  }
}

.sun-input-inner {
  width: 100%;
  padding: @input-padding-y @input-padding-x;
  border: @input-border-width solid @input-border-color;
  background-clip: padding-box;
  background-color: @input-bg;

  // Note: This has no effect on <select>s in some browsers, due to the limited stylability of `<select>`s in CSS.
  border-radius: @input-border-radius;
  box-shadow: @input-box-shadow;
  color: @input-color;
  font-family: @input-font-family;
  font-size: @input-font-size;
  font-weight: @input-font-weight;
  line-height: @input-line-height;
  transition: @input-transition;

  &:focus {
    border-color: @input-focus-border-color;
    background-color: @input-focus-bg;
    box-shadow: @input-focus-box-shadow;
    color: @input-focus-color;
    outline: 0;
    // Avoid using mixin so we can pass custom focus shadow properly
  }

  &::placeholder {
    color: @input-placeholder-color;
    // Override Firefox's unusual default opacity; see https://github.com/twbs/bootstrap/pull/11526.
    opacity: 1;
  }

  &:disabled,
  &[readonly] {
    border-color: @input-disabled-border-color;
    background-color: @input-disabled-bg;
    // iOS fix for unreadable disabled content; see https://github.com/twbs/bootstrap/issues/11655.
    opacity: 1;
  }
}

.icon-wrap + .sun-input-inner {
  padding-right: 35px;
}

.sun-input-group-prepend,
.sun-input-group-append {
  display: flex;
  align-items: center;
  padding: @input-padding-y @input-padding-x;
  border: @input-border-width solid @input-group-addon-border-color;
  margin-bottom: 0; // Allow use of <label> elements by overriding our default margin-bottom
  background-color: @input-group-addon-bg;
  border-radius: @input-border-radius;
  color: @input-group-addon-color;
  font-size: @input-font-size; // Match inputs
  font-weight: @font-weight-normal;
  line-height: @input-line-height;
  text-align: center;
  white-space: nowrap;
}

.input-size-sm .sun-input-inner {
  padding: @input-padding-y-sm @input-padding-x-sm;
  border-radius: @input-border-radius-sm;
  font-size: @input-font-size-sm;
}

.input-size-lg .sun-input-inner {
  padding: @input-padding-y-lg @input-padding-x-lg;
  border-radius: @input-border-radius-lg;
  font-size: @input-font-size-lg;
}

.sun-input-group-append + .btn {
  padding: 0;
  border: 0;
}

.input-group > .sun-input-group-prepend,
.input-group.input-group-append > .sun-input-inner {
  .border-right-radius(0);
}

.input-group > .sun-input-group-append,
.input-group.input-group-prepend > .sun-input-inner {
  .border-left-radius(0);
}
