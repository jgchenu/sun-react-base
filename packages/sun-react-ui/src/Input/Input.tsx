import React, { ReactNode, InputHTMLAttributes, ChangeEvent } from 'react';
import classNames from 'classnames';
import { sunPrefix } from '../common';
import './style.less';

type InputSize = 'large' | 'small';

interface BasicInputProps {
  /**是否禁用 Input */
  disabled?: boolean;
  /**设置 input 大小，支持 lg 或者是 sm */
  size?: InputSize;
  /**添加图标，在右侧悬浮添加一个图标，用于提示 */
  icon?: ReactNode;
  /**添加前缀 用于配置一些固定组合 */
  prepend?: string | ReactNode;
  /**添加后缀 用于配置一些固定组合 */
  append?: string | ReactNode;
  onChange?: (e: ChangeEvent<HTMLInputElement>) => void;
}

type InputProps = BasicInputProps &
  Omit<InputHTMLAttributes<HTMLInputElement>, 'size' | 'onChange' | 'disabled'>;

const inputPrefixClass = `${sunPrefix}-input`;

function Input(props: InputProps) {
  const { disabled, size, icon, prepend, append, style, ...restProps } = props;
  const classes = classNames(`${inputPrefixClass}-wrap`, {
    [`${inputPrefixClass}-${size}-size`]: !!size,
    [`${inputPrefixClass}-append-wrap`]: !!append,
    [`${inputPrefixClass}-prepend-wrap`]: !!prepend,
  });

  return (
    <div className={classes} style={style}>
      {prepend && (
        <div className={`${inputPrefixClass}-prepend`}>{prepend}</div>
      )}
      {icon && <div className={`${inputPrefixClass}-icon-wrap`}>{icon}</div>}
      <input
        className={`${inputPrefixClass}-inner`}
        disabled={disabled}
        {...restProps}
      />
      {append && <div className={`${inputPrefixClass}-append`}>{append}</div>}
    </div>
  );
}

export default Input;
