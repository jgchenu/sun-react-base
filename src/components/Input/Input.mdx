---
name: Input
route: /input
---

import "../../styles/index.less";
import Input from "./Input";
import { FaSearch } from "react-icons/fa";
import { useState } from "react";
import { Playground, Props } from "docz";

# Input

<Props of={Input} />

## Default Input

<Playground>
  <Input
    style={{ width: 200 }}
    placeholder="placeholder"
    onChange={console.log}
  />
</Playground>

## Control Input

<Playground>
  {() => {
    const [value, setValue] = useState();
    return (
      <Input
        value={value}
        defaultValue={value}
        onChange={(e) => {
          setValue(e.target.value);
        }}
      />
    );
  }}
</Playground>

## Disable Input

<Playground>
  <Input style={{ width: 300 }} placeholder="disabled input" disabled />
</Playground>

## Icon Input

<Playground>
  <Input
    style={{ width: 300 }}
    placeholder="input with icon"
    icon={<FaSearch />}
  />
</Playground>

## Pand Input

<Playground>
  <Input
    style={{ width: 300 }}
    defaultValue="prepend text"
    prepend="https://"
  />
  <Input style={{ width: "300px" }} defaultValue="google" append=".com" />
</Playground>

## Size Input

<Playground>
  <Input style={{ width: 300 }} defaultValue="large size" size="lg" />
  <Input style={{ width: 300 }} placeholder="small size" size="sm" />
</Playground>
