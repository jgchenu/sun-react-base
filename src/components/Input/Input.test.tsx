import React from "react";
import renderer from "react-test-renderer";

import Input, { InputProps } from "./Input";

const defaultProps: InputProps = {
  onChange: jest.fn(),
  placeholder: "test-input",
};
describe("test Input component", () => {
  it("Button should render correct html", () => {
    const tree = renderer.create(<Input {...defaultProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  // it("should render the correct default Input", () => {
  //   const wrapper = render(<Input {...defaultProps} />);
  //   const testNode = wrapper.getByPlaceholderText(
  //     "test-input"
  //   ) as HTMLInputElement;
  //   expect(testNode).toBeInTheDocument();
  //   expect(testNode).toHaveClass("sun-input-inner");
  //   fireEvent.change(testNode, { target: { value: "23" } });
  //   expect(defaultProps.onChange).toHaveBeenCalled();
  //   expect(testNode.value).toEqual("23");
  // });
  // it("should render the disabled Input on disabled property", () => {
  //   const wrapper = render(<Input disabled placeholder="disabled" />);
  //   const testNode = wrapper.getByPlaceholderText(
  //     "disabled"
  //   ) as HTMLInputElement;
  //   expect(testNode.disabled).toBeTruthy();
  // });
  // it("should render different input sizes on size property", () => {
  //   const wrapper = render(<Input placeholder="sizes" size="lg" />);
  //   const testContainer = wrapper.container.querySelector(".sun-input-wrap");
  //   expect(testContainer).toHaveClass("input-size-lg");
  // });
  // it("should render prepand and append element on prepand/append property", () => {
  //   const { queryByText, container } = render(
  //     <Input placeholder="pend" prepend="https://" append=".com" />
  //   );
  //   const testContainer = container.querySelector(".sun-input-wrap");
  //   expect(testContainer).toHaveClass(
  //     "input-group input-group-append input-group-prepend"
  //   );
  //   expect(queryByText("https://")).toBeInTheDocument();
  //   expect(queryByText(".com")).toBeInTheDocument();
  // });
});
