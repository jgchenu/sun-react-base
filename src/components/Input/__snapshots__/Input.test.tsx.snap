// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Input component Button should render correct html 1`] = `
<div
  className="sun-input-wrap"
>
  <input
    className="sun-input-inner"
    onChange={[MockFunction]}
    placeholder="test-input"
  />
</div>
`;
